[PlayStation] RunLoopGeneric has memory leak in vector of ScheduledTask

basuke opened the pull request:
#### cef910fdba391af2787482bc0b1b89c7b8a082d4
<pre>
[PlayStation] RunLoopGeneric has memory leak in vector of ScheduledTask
<a href="https://bugs.webkit.org/show_bug.cgi?id=241736">https://bugs.webkit.org/show_bug.cgi?id=241736</a>
&lt;rdar://problem/95890501&gt;

Reviewed by Yusuke Suzuki.

The ScheduledTask object is ref counted. There are two main owners, RunLoop::TimerBase m_scheduledTask
and RunLoop::m_schedules vector. When new timer is scheduled while current one is active but not fired,
it is marked as inactive to make it not fired in the future, and new one is created with active state.
The old one will be removed eventually when original timing come to fire.

On the other hand when JSRunLoopTimer is cancelled, it just set a new timer to WTF::RunLoop with the
long interval (it is 10 years) to cancel the running one. Then invalidated timers with those long interval
has no chance to be removed from the task list of WTF::RunLoop. With a simple application with using
repeating timer, it is easily observed that we see tens of thousands of inactive objects in the list in
just a few minutes.

In this patch, stop leaving the cancelled ScheduledTask as deactivated in m_schedules and remove it before
scheduling new one. To achieve this goal, the data structure of m_schedules was changed from heap to
RedBlackTree because it&apos;s hard to find the object in heap (that&apos;s just linear search).

Also TimerBase now always have one ScheduledTask in its lifecycle so that RefPtr was changed to Ref.
m_schedules keeps the pointer of ScheduledTask instead of Ref&lt;SheduledTask&gt; because there&apos;s no chance
to keep ScheduledTask in the tree while TimerBase is destructed. But it&apos;s still ref counted object because
there&apos;re timing while populateTasks returns the tasks while timer was gone.

* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/generic/RunLoopGeneric.cpp:
(WTF::RunLoop::TimerBase::ScheduledTask::create):
(WTF::RunLoop::TimerBase::ScheduledTask::ScheduledTask):
(WTF::RunLoop::TimerBase::ScheduledTask::fired):
(WTF::RunLoop::TimerBase::ScheduledTask::updateReadyTime):
(WTF::RunLoop::TimerBase::ScheduledTask::key const):
(WTF::RunLoop::TimerBase::ScheduledTask::isScheduled const):
(WTF::RunLoop::TimerBase::ScheduledTask::setScheduled):
(WTF::RunLoop::TimerBase::ScheduledTask::activate):
(WTF::RunLoop::populateTasks):
(WTF::RunLoop::runImpl):
(WTF::RunLoop::scheduleWithLock):
(WTF::RunLoop::unscheduleWithLock):
(WTF::RunLoop::TimerBase::TimerBase):
(WTF::RunLoop::TimerBase::start):
(WTF::RunLoop::TimerBase::stopWithLock):
(WTF::RunLoop::TimerBase::isActiveWithLock const):

Canonical link: <a href="https://commits.webkit.org/253065@main">https://commits.webkit.org/253065@main</a>
</pre>

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -156,25 +176,50 @@ void RunLoop::runImpl(RunMode runMode)
+            });
+            reporter->startRepeating(30_s);
+            fprintf(stderr, "[RunLoop][pid %d main %s] Reporter installed\n",
+                getpid(),
```
Let's use `ProcessID.h`'s function.

Constellation replied:
Ditto, let's use `dataLogLn` or `dataLogF`.

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -27,38 +27,37 @@
 #include <wtf/RunLoop.h>
 
+#include <unistd.h>
+
```
Let's drop it to keep it generic. ProcessID.h offers the platform-generic way to get process-id. (`unistd.h` is not generic header).

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -156,25 +176,50 @@ void RunLoop::runImpl(RunMode runMode)
+                    getpid(),
+                    (isMain() ? "true" : "false"),
+                    (int) m_schedules.size(),
+                    (int) active);
```
Let's use `dataLogLn`, or `dataLogF`.

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -234,29 +279,27 @@ RunLoop::CycleResult RunLoop::cycle(RunLoopMode)
 {
-    m_schedules.append(task.ptr());
-    std::push_heap(m_schedules.begin(), m_schedules.end(), TimerBase::ScheduledTask::EarliestSchedule());
+    return m_schedules.findIf(TimerBase::ScheduledTask::IdenticalSchedule(task)) != notFound;
```
This looks like `O(N)` operation. Can we keep `O(logN)`?
One possible implementation I can think is that, instead of using Vector for m_schedules, making it `RedBlackTree` (see `wtf/RedBlackTree.h`).
We can do that, so long as we do not modify `scheduledTimePoint` while it is in this tree. (And I think this is right. We are modifying it only after removing the node from this vector).

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -68,35 +70,52 @@ WTF_MAKE_NONCOPYABLE(ScheduledTask);
-            return lhs->scheduledTimePoint() > rhs->scheduledTimePoint();
-        }
-    };
+    MonotonicTime key() const {
```
Newline is required before `{`.

basuke replied:
Right.

Constellation commented on Source/WTF/wtf/generic/RunLoopGeneric.cpp:
```diff
@@ -270,18 +316,15 @@ void RunLoop::TimerBase::start(Seconds interval, bool repeating)
-    }, interval, repeating);
-    m_runLoop->scheduleAndWakeUpWithLock(*m_scheduledTask);
+    m_scheduledTask->activate(interval, repeating);
+    m_runLoop->scheduleWithLock(Ref(m_scheduledTask));
```
Do we need `Ref`?

basuke replied:
No! Thanks.

Constellation reviewed (approved):
r=me

basuke committed:
[PlayStation] RunLoopGeneric has memory leak in vector of ScheduledTask
https://bugs.webkit.org/show_bug.cgi?id=241736
<rdar://problem/95890501>

Reviewed by Yusuke Suzuki.

The ScheduledTask object is ref counted. There are two main owners, RunLoop::TimerBase m_scheduledTask
and RunLoop::m_schedules vector. When new timer is scheduled while current one is active but not fired,
it is marked as inactive to make it not fired in the future, and new one is created with active state.
The old one will be removed eventually when original timing come to fire.

On the other hand when JSRunLoopTimer is cancelled, it just set a new timer to WTF::RunLoop with the
long interval (it is 10 years) to cancel the running one. Then invalidated timers with those long interval
has no chance to be removed from the task list of WTF::RunLoop. With a simple application with using
repeating timer, it is easily observed that we see tens of thousands of inactive objects in the list in
just a few minutes.

In this patch, stop leaving the cancelled ScheduledTask as deactivated in m_schedules and remove it before
scheduling new one. To achieve this goal, the data structure of m_schedules was changed from heap to
RedBlackTree because it's hard to find the object in heap (that's just linear search).

Also TimerBase now always have one ScheduledTask in its lifecycle so that RefPtr was changed to Ref.
m_schedules keeps the pointer of ScheduledTask instead of Ref<SheduledTask> because there's no chance
to keep ScheduledTask in the tree while TimerBase is destructed. But it's still ref counted object because
there're timing while populateTasks returns the tasks while timer was gone.

* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/generic/RunLoopGeneric.cpp:
(WTF::RunLoop::TimerBase::ScheduledTask::create):
(WTF::RunLoop::TimerBase::ScheduledTask::ScheduledTask):
(WTF::RunLoop::TimerBase::ScheduledTask::fired):
(WTF::RunLoop::TimerBase::ScheduledTask::updateReadyTime):
(WTF::RunLoop::TimerBase::ScheduledTask::key const):
(WTF::RunLoop::TimerBase::ScheduledTask::isScheduled const):
(WTF::RunLoop::TimerBase::ScheduledTask::setScheduled):
(WTF::RunLoop::TimerBase::ScheduledTask::activate):
(WTF::RunLoop::populateTasks):
(WTF::RunLoop::runImpl):
(WTF::RunLoop::scheduleWithLock):
(WTF::RunLoop::unscheduleWithLock):
(WTF::RunLoop::TimerBase::TimerBase):
(WTF::RunLoop::TimerBase::start):
(WTF::RunLoop::TimerBase::stopWithLock):
(WTF::RunLoop::TimerBase::isActiveWithLock const):

Canonical link: https://commits.webkit.org/253065@main
